File: PGBooks/PGBooks/Sources/UI/ForYou/ForYouTopController.m
//
//  ForYouTopController.m
//  PGBooks
//
//  Created by ipagong on 11/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

#import "ForYouTopController.h"
#import "BookSimpleCell.h"
#import "UserDataService.h"

@interface ForYouTopController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) IBOutlet UITableView *tableview;

@property (nonatomic, strong) NSArray<PGBook*>* data;

@end

@implementation ForYouTopController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setupUI];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self dataLoad];
}

#pragma mark - tableview delegate & datasoruces

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.data.count;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    BookSimpleCell * cell = [tableView dequeueReusableCellWithIdentifier:[BookSimpleCell identifier] forIndexPath:indexPath];
    [cell bindData:[self.data objectAtIndex:indexPath.row]];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    PGBook *book = [self.data objectAtIndex:indexPath.row];
    
    [BookDetailController showWithBook:book target:self];
}

#pragma mark -

- (void)setupUI
{
    self.title = @"Favorite";
    
    self.tableview.delegate   = self;
    self.tableview.dataSource = self;
    self.tableview.tableFooterView = [UIView new];
    
    [BookSimpleCell registOnTableView:self.tableview];
}

- (void)dataLoad {
    self.data = [UserDataService favoriteBooks];
    
    [self.tableview reloadData];
}

@end

File: PGBooks-swift/PGBooks-swift/Sources/UI/ForYou/ForYouTopController.swift
//
//  ForYouTopController.swift
//  PGBooks-swift
//
//  Created by ipagong on 16/10/2019.
//  Copyright © 2019 ipagong. All rights reserved.
//

import UIKit
import RxSwift
import RxCocoa
import RxDataSources

class ForYouTopController: UIViewController, BookDetailTransitionable {

    @IBOutlet weak var tableview : UITableView!
    
    private lazy var viewModel:ForYouViewModel = { ForYouViewModel.init() }()
    private let disposeBag = DisposeBag()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        binding()
    
    }
    
    func setupUI() {
        self.title = ServiceType.forYou.title
        self.tableview.register(BookSimpleCell.nibByIdentifier,
                                forCellReuseIdentifier: BookSimpleCell.identifier)
        self.tableview.tableFooterView = UIView()
    }
    
    func binding() {
        self.lifeCycle.signalViewWillAppear()
            .bind(to: self.viewModel.input.loadSignal)
            .disposed(by: self.disposeBag)
        
        self.viewModel.output.sections
            .bind(to: self.tableview.rx.items(dataSource:
                RxTableViewSectionedReloadDataSource<Book.Section>.init(configureCell: { (ds, tb, ip, data) -> UITableViewCell in
                    return tb.getCell(value: BookSimpleCell.self, indexPath: ip, data: data)
                })
            ))
            .disposed(by: self.disposeBag)
        
        self.viewModel.output.sections
            .flatMapLatest{ [weak self] data in
                self?.tableview.rx.itemSelected.map{ data[$0.section].items[$0.row] } ?? .empty()
            }.asOptional()
            .catchErrorJustReturn(nil).unwrap()
            .subscribe(onNext: { [weak self] (book) in
                guard let self = self else { return }
                self.showBookDetail(book)
            }).disposed(by: self.disposeBag)
    }

}
